src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Semantics/Symbols/SymbolManagerBase.cs

```diff
@@ -328,26 +328,34 @@ public TypeArrayKey(CType[] types)
 
             public bool Equals(TypeArrayKey other)
             {
-                if (other._types == _types) return true;
-                if (other._types.Length != _types.Length) return false;
-                if (other._hashCode != _hashCode) return false;
-                for (int i = 0; i < _types.Length; i++)
+                CType[] types = _types;
+                CType[] otherTypes = other._types;
+                if (other._types == _types)
                 {
-                    if (!_types[i].Equals(other._types[i]))
+                    return true;
+                }
+
+                if (other._hashCode != _hashCode || otherTypes.Length != types.Length)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < otherTypes.Length; i++)
+                {
+                    if (!types[i].Equals(otherTypes[i]))
+                    {
                         return false;
+                    }
                 }
+
                 return true;
             }
 
             [ExcludeFromCodeCoverage] // Typed overload should always be the method called.
             public override bool Equals(object obj)
             {
                 Debug.Fail("Sub-optimal overload called. Check if this can be avoided.");
-                if (obj is TypeArrayKey)
-                {
-                    return this.Equals((TypeArrayKey)obj);
-                }
-                return false;
+                return obj is TypeArrayKey && Equals((TypeArrayKey)obj);
```

Comment:
Is that known to jit to something nicer? The IL is a bit bulkier, with the closest C# 6 being:

```
TypeArrayKey? temp = obj as TypeArrayKey?;
TypeArrayKey k = obj.GetValueOrDefault();
return temp.HasValue && Equals(k);
```

If I was going to do more with the `k` than a single call, I might still favour it for the way the variable gets scoped, but that doesn't seem like a readability gain to me either when I'm just doing that one further call. Maybe once I'm more used to it I'll think differently (I hated expression-body members when they first came out and find them much clearer for single-expression members now).